File: leetcode/medium/binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    vector<vector<int>> ans;
public:
    void traversal(TreeNode* node, int level) {
        if(node == NULL) {
            return;
        }
        
        if(ans.size() < level + 1) {
            ans.push_back({});
        }
        ans[level].push_back(node -> val);
        
        traversal(node -> left, level + 1);
        traversal(node -> right, level + 1);
    }
    
    void swap(int i) {
        int right = ans[i].size() - 1;
        int left = 0;
        
        while(left < right) {
            int tmp = ans[i][left];
            ans[i][left] = ans[i][right];
            ans[i][right] = tmp;
            
            left++;
            right--;
        }
    }
    
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        traversal(root, 0);
        
        for(int i = 0; i < ans.size(); i++) {
            if(i % 2 != 0) {
                swap(i);
            }
        }
        
        return ans;
    }
};

File: leetcode/medium/binary-tree-zigzag-level-order-traversal.py
from collections import defaultdict, deque

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        ans = defaultdict(list)
        q = deque([(root, False, 0)])
        
        while q:
            node, flag, lvl = q.popleft()
            if lvl % 2 == 1:
                ans[lvl].insert(0, node.val)
            else:
                ans[lvl].append(node.val)
            childs = [node.left, node.right]
            for child in childs:
                if not child:
                    continue
                q.append((child, not flag, lvl + 1))
        return ans.values()
        
